platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/CompoundAssignmentResult.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file.collections;
+
+import com.google.common.base.Preconditions;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.internal.file.UnionFileCollection;
+import org.gradle.api.internal.provider.support.SupportsCompoundAssignment;
+import org.gradle.api.internal.tasks.TaskDependencyFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * A helper class to implement an intermediate result of a compound assignment operation, like "+=".
+ * It is then assigned to the left-hand side operand. When the LHS is a ConfigurableFileCollection-typed property of some Gradle-enhanced object, then the assignment action is invoked.
+ */
+class CompoundAssignmentResult extends UnionFileCollection implements SupportsCompoundAssignment.Result<FileCollection> {
+    @Nullable
+    private ConfigurableFileCollection owner;
+    private final FileCollectionInternal rhs;
+
+    public CompoundAssignmentResult(TaskDependencyFactory taskDependencyFactory, DefaultConfigurableFileCollection owner, FileCollectionInternal rhs) {
+        super(taskDependencyFactory, owner, rhs);
+        this.owner = owner;
+        this.rhs = rhs;
+    }
+
+    public boolean isOwnedBy(ConfigurableFileCollection owner) {
+        return this.owner == owner;
+    }
+
+    public void assignToOwner() {
+        ConfigurableFileCollection theOwner = owner;
+        owner = null;
+        Preconditions.checkState(theOwner != null, "The collection is already consumed by the owner");
```

Comment:
❔ Should we add a similar check to `unwrap` to catch cases of accidental double-unwraps?